Whatsnew fixes

pp-mo opened the pull request:
Tighten up the naming format for whatsnew contributions files, and add a master test to ensure that the contents of the latest contributions directory all comply.

Problems encountered on the way to using this to define 1.9 release notes.
Before this, it was failing on some files because :
 * (1)  one file was named with year-day-month instead of year-month-day : a "month" > 12 gave an error.
 * (2) one file was named without a summary title, which similarly failed.

Hopefully I fixed those right, but some guesswork was involved, hence this attempt to tighten things up.
It now fails tests if you provide a contributions file with an badly formatted name.

As a bonus, you can now run aggregate_directory without specifying a version :  it finds the latest and builds that.

**Testing Notes:**
tested with : badly named files ; no files ; no "contributions_*" directory

rhattersley commented on docs/iris/src/whatsnew/aggregate_directory.py:
```diff
@@ -176,9 +260,16 @@ def run_compilation(release):
     PARSER = argparse.ArgumentParser()
+    _DUMMY_VERSION = '9999.9999'
     PARSER.add_argument("release", help="Release number to be compiled",
+                        nargs='?', default=_DUMMY_VERSION,
```
You don't need to use a hacky/dummy version ... you can leave this off and argparse will happily give you a `None` if the argument was missing.

rhattersley commented on docs/iris/src/whatsnew/aggregate_directory.py:
```diff
@@ -165,9 +235,23 @@ def compile_directory(directory, release):
 
 
-def run_compilation(release):
+def check_all_contributions_valid(release=None):
```
This routine isn't used??

pp-mo replied:
It is used, by the new test.
I could move it there, but I'd rather not as I'd need to make some more bits public, and even then I think it is using implementation-specific assumptions.

rhattersley replied:
> It is used, by the new test.

That never occured to me! :expressionless:

rhattersley commented on docs/iris/src/whatsnew/aggregate_directory.py:
```diff
@@ -176,9 +260,16 @@ def run_compilation(release):
+    release = ARGUMENTS.release
+    if release == _DUMMY_VERSION:
+        release = get_latest_release()
+    print('Building release document for release "{!s}"'.format(release))
```
Given you've made the `release` argument for `run_compilation()` optional it would be simpler to move this print-out inside that routine. Then you could make the `release` argument of `find_release_directory` mandatory, which "tightens up" your API and makes it less likely to be used incorrectly in future due to a missing argument. It also reduces redundant behaviour and improves the single-point-of-responsibility - i.e. figure out which release you want in one place, then re-use that information.

pp-mo replied:
There's a lot I'm not keen on in the control/calling structure of this module, but I didn't want to  rewrite more than I had to, as I felt the reworks were getting unnecessarily involved and I might encounter opposition to spending so much effort on this non-functional code (i.e. including the reviewers' !).

If I were to review + refactor the whole thing, I would certainly want to rationalise some more things, like the way directory + release are passed around.

Even with this much refactoring and testing, I was already itching to add written tests for this, but I feared that might also be judged excessive !

rhattersley commented on lib/iris/tests/test_whatsnew_contributions.py:
```diff
@@ -0,0 +1,65 @@
+            sys.path = old_paths
+
+        # Run the check routine.
+        aggregate_directory.check_all_contributions_valid()
```
Fiddling with the path strikes me as a messy and potentially "dangerous" way of doing things. In a multi-threaded context other threads will now be subject to a modified path, and if they were equally badly behaved the logic to reset the path could trash their modifications. End result: intermittent weirdness! How about having a "validate" mode for the command line instead?

```
python aggregate_directory.py -c
```

rhattersley replied:
NB.

1. A check/validate mode would be handy for contributors anyway.
2. It would avoid the need to expose code in aggregate_directory which is only used in the tests.

pp-mo replied:
> How about having a "validate" mode for the command line instead?

That might be generally useful anyway, but I don't really much like calling out to the command line within tests.
Do you maybe know another way, perhaps involving the import mechanisms?

rhattersley replied:
`runpy.run_path`?

pp-mo replied:
Nope, it seems that is equivalent to execfile instead.
So it doesn't build a module, and the resulting definitions don't share a common global space for mutual access. 
I've gone with subprocess instead :-[

pp-mo commented:
At least I got rid of the path-twiddling.
This may be better squashing up now.

rhattersley commented on lib/iris/tests/test_whatsnew_contributions.py:
```diff
@@ -0,0 +1,63 @@
+            cmd = 'python aggregate_directory.py --checkonly --quiet'
+            subprocess.check_call(cmd, shell=True)
+        finally:
+            os.chdir(old_cwd)
```
You don't need the try-finally stuff, you can just use `subprocess.check_call(..., cwd=whatsnew_dirpath)`

rhattersley replied:
(The Python documentation for that is a little obscure, so I can see why you might not have spotted it!)

pp-mo replied:
Thanks, I'll do that..

pp-mo commented:
That didn't go so well.
I assume the problem is that paths with '..' elements don't work here, since the previous code did work, so I've reverted that logic.  
We'll see whether that works here.

I also used check_call(cwd=...) as suggested.

pp-mo commented:
**NOTE**:
@rhattersley + whoever
I'm playing with this repo fixing a Travis problem.
Please ignore till I get this sorted :  I will be getting rid of junk commits with a rebase, later.

pp-mo commented:
Hi @rhattersley and whoever

Now fixed the testing problem.
Note that to get Travis to test this I needed a separate invocation from .travis.yml, as Iris is running in a python install so it doesn't know where the repo/build directories are.

I also rebased as it had got very complicated, and I needed to include + rename subsequent contributions entries to get it to pass.

I think all the outstanding points raised are now addressed.

rhattersley commented:
You also need to update `docs/iris/src/developers_guide/documenting/whats_new_contributions.rst` to mention the new date format.

pp-mo commented:
> You also need to update docs/iris/src/developers_guide/documenting/whats_new_contributions.rst to mention the new date format.

Thanks, I hadn't actually realised there was a description elsewhere !

pp-mo commented:
>  update docs/iris/src/developers_guide/documenting/whats_new_contributions.rst 

I've now tweaked that a bit.  I hope it makes sense.

pp-mo commented:
**NOTE: don't merge this**
We should merge this into the new 1.9.x branch **first** (it will help me with 1.9 release)
So, let me know when you're happy to merge + I'll do it via a re-targetted PR

rhattersley commented:
> So, let me know when you're happy to merge + I'll do it via a re-targetted PR

:+1: Go for it!

pp-mo commented:
Replaced by #1854
